llvm/lib/Target/BPF/BPFCheckUndefIR.cpp

```diff
@@ -0,0 +1,118 @@
+//===---------------- BPFAdjustOpt.cpp - Adjust Optimization --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Check 'undef' and 'unreachable' IRs and issue proper warnings.
+//
+//===----------------------------------------------------------------------===//
+
+#include "BPF.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Type.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Pass.h"
+
+#define DEBUG_TYPE "bpf-check-undef-ir"
+
+using namespace llvm;
+
+namespace {
+
+class BPFCheckUndefIR final : public ModulePass {
+  bool runOnModule(Module &F) override;
+
+public:
+  static char ID;
+  BPFCheckUndefIR() : ModulePass(ID) {}
+
+private:
+  void BPFCheckUndefIRImpl(Function &F);
+  void BPFCheckInst(Function &F, BasicBlock &BB, Instruction &I);
+  void HandleReturnInsn(Function &F, ReturnInst *I);
+  void HandleUnreachableInsn(Function &F, BasicBlock &BB, Instruction &I);
+};
+} // End anonymous namespace
+
+char BPFCheckUndefIR::ID = 0;
+INITIALIZE_PASS(BPFCheckUndefIR, DEBUG_TYPE, "BPF Check Undef IRs", false,
+                false)
+
+ModulePass *llvm::createBPFCheckUndefIR() { return new BPFCheckUndefIR(); }
+
+void BPFCheckUndefIR::HandleReturnInsn(Function &F, ReturnInst *I) {
+  Value *RetValue = I->getReturnValue();
+  // PoisonValue is a special UndefValue where compiler intentionally to
+  // poisons a value since it shouldn't be used.
+  if (!RetValue || isa<PoisonValue>(RetValue) || !isa<UndefValue>(RetValue))
```

Comment:
Looks like undef is deprecated.
Do we have to check it or PoisonValue is enough?

Reply:
For some extreme case, e.g., 
```
int foo(void) {
  int i[2];
  return i[1];
}
```
The eventual IR will look like
```
define dso_local i32 @foo() #0 {
  ret i32 undef
}
```
For such cases, checking `!isa<PoisonValue>(RetValue)` is not enough, we should check `!isa<UndefValue>(RetValue)` to ensure to report something wrong to user.

But such case should be really rare and it is easier for user to find out what is going on.

The prog Marc Suñé reported does not need the above check. Since upstream doesn't like to involve 'undef', I will remove the above checking and also remove the simple test which results in IR 'ret i32 undef' in the next revision.